Enable videos autoplay

miankhalid opened the pull request:
https://openedx.atlassian.net/browse/MA-1683

- Auto play enabled behind a flag named IS_AUTOPLAY_ENABLED
- Progress wheel now shows until the video has loaded
- Time variables standardized

plz review @1zaman @bguertin

1zaman commented:
Looks like this has a side-effect causing the video to not be paused when the Activity is paused.

bguertin commented:
If a user pauses a video and then pauses the activity (e.g. by going to home screen), then I don't think the video should auto-play when they go back to the app; it should remain paused.

Could we preserve the pause state so it only auto-plays the video if the user hasn't just paused it?

bguertin commented:
Besides the pause / auto-resume issue, I think this looks pretty good.

miankhalid commented:
@bguertin good point. Working on it.

miankhalid commented:
another pass ? @bguertin @1zaman

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -213,43 +211,41 @@ public void onResume() {
+     * @param play     Boolean to set state to playing or paused
+     * @param callback The callback to send events to
+     */
+    private void setVideoPlayerState(boolean play, IPlayerEventCallback callback) {
```
super nit: `play` -> `playing`

BenjiLee commented:
This is more of a visual nit, when the video is loading, the progress spinner does not consistently should up. It either shows the progress spinner, or the grayed out video player chrome.

Otherwise LGTM

miankhalid commented:
>This is more of a visual nit, when the video is loading, the progress spinner does not consistently show up. It either shows the progress spinner, or the grayed out video player chrome.

@BenjiLee I thought I had fixed this issue, but somehow it was still there. Nevertheless, I've fixed it now for good :wink: and have amended my previous commit with the super nit as well.

@bguertin if the video is paused and we come back from background (or next/previous page), the video remains paused and we show the video controller, otherwise, we play the video and show the video controller as well.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -486,9 +491,6 @@ public synchronized void playOrPrepare(String path, int seekTo, String title,
         }
 
-        // clear all errors
-        clearAllErrors();
```
Should we have `hideProgress()` separated from `clearAllErrors()`, and still clear just errors here? We can have a new `clearState()` method that calls both, and replace other calls to `clearAllErrors()` with it.

miankhalid replied:
This call here was redundant and clearing errors here will still be redundant, as all the errors are cleared once we go into `onPrepared`, which is actually the right place to do it.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -750,26 +752,27 @@ private void clearAllErrors() {
         }
 
+        environment.getSegment().trackVideoLoading(videoEntry.videoId, videoEntry.eid,
+                videoEntry.lmsUrl);
```
Why move this before `clearAllErrors()`? The tracking call should probably be performed at the end.

miankhalid replied:
ok

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -750,26 +752,27 @@ private void clearAllErrors() {
-        // mark prepared and allow orientation
-        isPrepared = true;
         initializeClosedCaptioning();
         allowSensorOrientation();
```
Maybe `allowSensorOrientation()` should be moved before the visibility check, so that it's set up in all cases.

miankhalid replied:
nope, we should only allow orientation changes when the video has loaded and is in playable state.

miankhalid replied:
good catch.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -975,15 +978,21 @@ private boolean isScreenLandscape() {
+                        player.start();
+                    } else {
+                        updateController("player paused");
+                    }
```
I think starting the player should not need to be performed from this delayed unfreeze callback; you should be able to start it directly from `onResume()`.

miankhalid replied:
there's a comment on one of the callbacks call which reads:
>// start playback after 300 milli seconds, so that it works on HTC One, Nexus5, S4, S5
 // some devices take little time to be ready

So this seems fine i guess.

1zaman replied:
Yeah so start and then freeze it? That should probably be fine, and seems to be how it was intended to be implemented.

1zaman replied:
I think the thing to do is actually just start the player in `onCreate()` once (not in `onResume()`. Freezing the player should already be handled, and you can remove the custom logic for it.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -1755,33 +1764,17 @@ public void callPlayerSeeked(long lastPostion, long newPosition, boolean isRewin
 
     /**
-     * Displays controller is in PORTRAIT MODE, otherwise hides controller.
+     * Displays controller based on a source
```
"based on a source" seems quite ambiguous; it can probably be removed.

miankhalid replied:
It really helps in debugging. We should keep it.

miankhalid replied:
or are you of the opinion that, we just remove "based on a source" from the description only ?

1zaman replied:
I mean just remove it from the method description.

miankhalid replied:
ok

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -1755,33 +1764,17 @@ public void callPlayerSeeked(long lastPostion, long newPosition, boolean isRewin
             // controller should also refresh, so hide and show it
             player.hideController();
-            
-            // if this is LANDSCAPE mode, then let controller be HIDDEN by default
```
Is this no longer relevant?

miankhalid replied:
yup not relevant. We should show the controller in any case.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -213,43 +211,41 @@ public void onResume() {
+    }
+
+    /**
+     * Sets the state of video player inside {@link PlayerFragment} to playing or paused
```
This could be changed to "Sets the playing/paused state of {@link PlayerFragment}" for concision.

miankhalid replied:
a `PlayerFragment` can never be in a playing/paused state, its the video player inside it that does these things. So, do we still want this description changed ?

1zaman replied:
I think it's fine, the main component in the Fragment is the video player. If you want you can add it though:

    Sets the playing/paused state of the video player in {@link PlayerFragment}

miankhalid replied:
ok

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -213,43 +211,41 @@ public void onResume() {
+    /**
+     * Sets the state of video player inside {@link PlayerFragment} to playing or paused
+     *
+     * @param playing  Boolean to set state to playing or paused
```
The description could perhaps be changed to "True for playing the video player, false for pausing it.", or something similar for more clarity.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -213,43 +211,41 @@ public void onResume() {
+     * Sets the state of video player inside {@link PlayerFragment} to playing or paused
+     *
+     * @param playing  Boolean to set state to playing or paused
+     * @param callback The callback to send events to
```
The callback parameter can be removed I guess, since it can be inferred from the value of the `playing` parameter.

miankhalid replied:
ok

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -350,25 +355,25 @@ public void onResume() {
     public void onPause() {
         super.onPause();
         if ( !isInViewPager ) {
+            if (player != null) isPaused = player.isPaused();
```
I think you can use freezing instead of pausing to have the player automatically save the playing state.

1zaman commented:
I commented on a few small issue, otherwise :+1:

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -975,15 +966,19 @@ private boolean isScreenLandscape() {
-                    if (pauseDueToDialog){
+                    if (IS_AUTOPLAY_ENABLED && isPrepared && !isAutoPlayDone) {
+                        isAutoPlayDone = true;
+                        player.start();
```
I think the auto-play logic should be performed in `onPrepared()` instead of from the unfreeze callback.

miankhalid replied:
It won't cater the case, when the video fragment isn't visible inside the view pager and `onPrepared()` is called.

1zaman replied:
Why not? There is a check around the start of the `onPrepared()` implementation, which returns from the method if `getUserVisibleHint()` returns false. If anything, the scenario isn't being handled in the current implementation.

1zaman replied:
OK, so I guess there is no way to change the video state without unfreezing it. I guess the current implementation is fine then. There should also be a check for `getUserVisibleHint()` at the start of the callback though, just like in `onPrepared()` (the callback is removed is `onPause()`, but `setUserVisibleHint()` is not called at all for child fragments so there is no way to have it removed from the event itself). Then this case should be handled completely.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -203,25 +207,25 @@ public void onComplete(FacebookDialog.PendingCall pendingCall, Bundle data) {
-                if(savedInstanceState.getBoolean("isMessageDisplayed")){
-                    showVideoNotAvailable(VideoNotPlayMessageType.IS_VIDEO_MESSAGE_DISPLAYED);
-                }
+                    && savedInstanceState.containsKey("player")) {
```
These saved state keys should be defined as constants.

1zaman replied:
Also the `containsKey()` check can be removed:

```` java
player = (IPlayer) savedInstanceState.get("player");
if (player != null) {
    player = new Player();
}
````

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -203,25 +207,25 @@ public void onComplete(FacebookDialog.PendingCall pendingCall, Bundle data) {
-                    player = new Player();
-                }
-            }
+        if (player == null) {
```
`player` is always null at this point, so this check is not needed.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -203,25 +207,25 @@ public void onComplete(FacebookDialog.PendingCall pendingCall, Bundle data) {
-        }catch(Exception e){
-            logger.error(e);
         }
+        if (savedInstanceState != null
```
Better to group all the logic inside one `null` check for `savedInstanceState`.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -203,25 +207,25 @@ public void onComplete(FacebookDialog.PendingCall pendingCall, Bundle data) {
-            logger.error(e);
         }
+        if (savedInstanceState != null
+                && savedInstanceState.containsKey("isMessageDisplayed")) {
```
As above, the `containsKey()` check is not necessary, as the `getBoolean()` method will return `false` if they key is not present.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -203,25 +207,25 @@ public void onComplete(FacebookDialog.PendingCall pendingCall, Bundle data) {
+            isPrepared = savedInstanceState.getBoolean("isPrepared");
+            isAutoPlayDone = savedInstanceState.getBoolean("isAutoPlayDone");
+        }
+        reAttachPlayEventListener();
```
This is already being called from `setupController()` (which is in turn called from `handleOnResume()`), so I think it can be removed from here.

1zaman commented:
A few small issues/nit-picks, otherwise :+1:

miankhalid commented:
@1zaman https://github.com/edx/edx-app-android/pull/457 caused some issues with play/pause persistence while we switched pages of the view-pager, which I've fixed in 543f008.

Wanna review ? :wink:

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/Player.java:
```diff
@@ -552,179 +551,185 @@ public void reset() {
-        }
+    @Override
+    public boolean isFrozen() {
+        return isFreeze;
```
Maybe have the variable renamed to `isFrozen` as well? It's more grammatically correct anyway :wink:

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -1800,6 +1783,15 @@ public VideoModel getPlayingVideo() {
     }
+
+    /**
+     * Returns true if video player is in freeze state
```
"freeze" can be replaced with "frozen" here as well I guess.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -288,8 +284,10 @@ public void onNegativeClicked() {
             } else {
-                //Video is downloaded. Hence play
-                startOnlinePlay(entry);
+                if (playerFragment != null && !playerFragment.isFrozen()) {
```
This could be converted into a `else if` I guess. Also, unless `setUserVisibleHint()` can be called before `onActivityCreated()`, `playerFragment` should always be initialized, so there is no need for the null check. It's performed in a lot of other places too though, so they can all be removed or it's fine as it is I guess

miankhalid replied:
`setUserVisibleHint()` is called before `onActivityCreated()`. So, null check is valid here.

1zaman commented:
Other than a few minor nitpicks :+1:

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -441,23 +446,17 @@ private void hideProgress() {
         }
+        outState.putBoolean(KEY_PREPARED, isPrepared);
+        outState.putBoolean(KEY_AUTOPLAY_DONE, isAutoPlayDone);
+        //FIXME: ensure that prepare is called on all activity restarts and then this can be removed
```
[MA-1820] (https://openedx.atlassian.net/browse/MA-1820) has been opened to address this issue.
